### PR TITLE
patch_manager: Prefer specific patches over global patches

### PR DESCRIPTION
**Pull request by elad335**
It mekes more sense that patches with specific game IDs will be preferred over global patches and not the opposite way arround like it was.
Memory contents works like LIFO: what's written last will be what will remain in memory.

### CODE REVIEW

**Comment by Megamouse** [comment]
This basically just breaks the intended behaviour of prefering game patches over global patches.
So you're doing the opposite of what you think you're doing.

See m_applied_groups for reference

**Comment by elad335** [comment]
I just tested it with 2 patches without "Group" node and it works as expected: all were applied, the first patch applied is with game ID, but for all versions. The second applied is also with same game ID, but for version 01.00. Both apply at the same address with different values and from looking at the debugger, the second patch's value is what it's in memory. 
Maybe you mean I should also take "Group" node into account?

**Comment by Megamouse** [comment]
Maybe first do one filter with the existing group logic and then simply revert the order for patch application

**Comment by elad335** [comment]
Fixed and optimized the code further.

File: Utilities/bin_patch.cpp [hunk 1]
```diff
@@ -705,52 +705,71 @@ std::basic_string<u32> patch_engine::apply(const std::string& name, u8* dst, u32
...
+			{
+				if (!m_applied_groups.emplace(patch->patch_group).second)
+				{
+					patch = nullptr;
+				}
+			}
+		}
+	}
+
+	// Sort specific patches after of global patches
```
Comment (Megamouse) [inline Utilities/bin_patch.cpp:762]:
```suggestion
	// Sort specific patches after global patches
```

File: Utilities/bin_patch.cpp [hunk 2]
```diff
@@ -705,52 +705,71 @@ std::basic_string<u32> patch_engine::apply(const std::string& name, u8* dst, u32
...
+	const auto patch_super_list =
+	{
+		&patches_for_all_serials_and_all_versions,
+		&patches_for_all_serials_and_this_version,
+		&patches_for_this_serial_and_all_versions,
+		&patches_for_this_serial_and_this_version
+	};
+
+	// Filter by patch group
+	for (auto it = std::rbegin(patch_super_list); it != std::rend(patch_super_list); it++)
```
Comment (Megamouse) [inline Utilities/bin_patch.cpp:748]: This still looks like you add global group patches first and then skip specific patches if they are in the same group.

Comment (elad335) [reply Utilities/bin_patch.cpp:748]: No it checks spefici patches first.

Comment (Megamouse) [reply Utilities/bin_patch.cpp:748]: Ah sorry. I didn't see the rbegin

File: Utilities/bin_patch.cpp [hunk 3]
```diff
@@ -705,52 +705,71 @@ std::basic_string<u32> patch_engine::apply(const std::string& name, u8* dst, u32
...
+			{
+				if (!m_applied_groups.emplace(patch->patch_group).second)
+				{
+					patch = nullptr;
+				}
+			}
+		}
+	}
+
+	// Sort specific patches after global patches
```
Comment (Megamouse) [inline Utilities/bin_patch.cpp:762]: This comment should be moved to where you actually sort them

Comment (elad335) [reply Utilities/bin_patch.cpp]: Fixed.

**Pull request merged**